### DISCUSSION TITLE
[PyTorch] Changing layers from torch.nn.functional to torch.nn affects attack accuracy.

### DISCUSSION THREAD

**Discussion by KarthikGanesan88** [Trusted-AI/adversarial-robustness-toolbox#1428] [Q&A]
Hi, I am using PyTorch with ART and if I swap the layers in my model from `torch.nn.functional` versions to the `torch.nn` versions, I am seeing big differences in attack accuracy. For example, in the [get_started_pytorch.py](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/examples/get_started_pytorch.py) file, when I run it as is, I get:

```
Accuracy on benign test examples: 98.24%
Accuracy on adversarial test examples: 40.62%
```
However, if I change the network to be this instead:

```
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv_1 = nn.Conv2d(in_channels=1, out_channels=4, kernel_size=5, stride=1)
        self.relu_1 = nn.ReLU(inplace=True)
        self.mp_1 = nn.MaxPool2d(2, 2)
        self.conv_2 = nn.Conv2d(in_channels=4, out_channels=10, kernel_size=5, stride=1)
        self.relu_2 = nn.ReLU(inplace=True)
        self.mp_2 = nn.MaxPool2d(2, 2)
        self.fc_1 = nn.Linear(in_features=4 * 4 * 10, out_features=100)
        self.relu_3 = nn.ReLU(inplace=True)
        self.fc_2 = nn.Linear(in_features=100, out_features=10)

    def forward(self, x):
        x = self.conv_1(x)
        x = self.relu_1(x)
        x = self.mp_1(x)
        x = self.conv_2(x)
        x = self.relu_2(x)
        x = self.mp_2(x)
        x = x.view(-1, 4 * 4 * 10)
        x = self.fc_1(x)
        x = self.relu_3(x)
        x = self.fc_2(x)
        return x
```
My results change to: 

```
Accuracy on benign test examples: 97.46%
Accuracy on adversarial test examples: 14.74%
```

As far as I am aware, both of these networks should work the exact same way, but PyTorch recommends using the `torch.nn` versions instead of calling the `torch.nn.functional` functions directly. Is there something in ART that requires `relu` and `maxpool2d` to be called as functions during forward but not have them as layers in the model? 

I also tried training both of these nets with PyTorch directly and got the same classification accuracy of `98%` each time. 

I am only showing the attack success change for `FGSM` but I see similar discrepancies for `PGD, HopSkipJump, CarliniL2 and SaliencyMap`.

**Comment by beat-buesser** [comment]
Hi @KarthikGanesan88 That's an interesting observation, but I don't think it is caused by your new model or ART. Running the original script multiple times usually results in adversarial accuracies around 35%, but sometimes results in values as low as 16% or as high as 50%. A possible reason could be that the model gets initialised and trained slightly different in each run. It could also be that the dataset MNIST is rather sensitive to small differences, maybe we should repeat the experiments with CIFAR10.
I think your new model is not exactly the same because it uses the inplace option for ReLU whereas the original model does not use it. It could be that this also introduces numerical differences that could affect the sign-based attacks, but I have not tested it in detail.

**Reply by beat-buesser** [reply]
One idea could be to initialise both model definitions with the same weights and biases loaded from disc.
Please let us know what you find out!

**Reply by KarthikGanesan88** [reply]
Hi @beat-buesser , I tried the experiments you suggested. Here is what I am seeing: 

1. Removing the `inplace=True` for `ReLU` made no difference. 
2. I used the same initial weights for training with PyTorch and with ART (by setting `torch.manual_seed(0)` at the start. I also printed the weights to confirm they are the same in both places, each time).
3. I also matched the criterion and optimizer as follows:
```
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.01)
```
4. Initially, I had PyTorch normalize the inputs in the range `(-1.0, 1.0)` while ART was doing `(0.0, 1.0)`. Now they are both doing `(0.0, 1.0)`

With these changes, PyTorch training is always giving me exactly 98% classification accuracy. I am also printing the `loss` every 2000 inputs for each epoch and they are identical during each training run. 

But ART still gives me (slightly) different classification accuracies, from `93.67%` to `94.77%`. So it seems there is still some inherent randomness when training with ART vs. with PyTorch? 

The adversarial attack numbers are also more stable now. The range is only from `13.77%` to `17.73%` for `FGSM`. 

Your last suggetsion of trying with `CIFAR-10` is my next step. I am training `AlexNet` on `CIFAR-10` now and will see how these observations translate to that dataset.

**Reply by beat-buesser** [reply]
Hi @KarthikGanesan88 Thank you very much for the detailed response! It's great to see that the ranges got narrower, but they should be identical to the one of the model. Would you be able to share the script that creates the numbers for MNIST with PyTorch and ART? I would be very interested to take a closer look.

**Reply by KarthikGanesan88** [reply]
Hi @beat-buesser , sure. Here are the files available as a [gist](https://gist.github.com/KarthikGanesan88/d5091e02fde3e9c93583d570a4488850). This is just for `MNIST`. My `CIFAR-10` model is still training!

**Reply by beat-buesser** [reply]
@KarthikGanesan88 Thank you! I think a big difference is that you are shuffling the training data with `torch.utils.data.DataLoader` whereas ART currently shuffles the data with Numpy. Can you try with setting Numpy's seed with
```python
import numpy as np
np.random.seed(0)
```

**Reply by KarthikGanesan88** [reply]
@beat-buesser No luck, I'm afraid. even with `np.random.seed(0)`, I am seeing the training result in accuracies from 93% to 96%, while I am seeing PyTorch training always yield 98%. I also tried setting `shuffle=False` in `torch.utils.data.DataLoader` but still got the same 98%. 

One thing though is that PyTorch takes about 5x longer to train than ART, even when things like number of training inputs, epochs, optimizer, loss function are identical. Not sure if that might be related to this issue.

**Reply by beat-buesser** [reply]
@KarthikGanesan88 I looked again at the code and noticed it uses `random`'s shuffle and not Numpy, can you try setting `random.seed`?

**Reply by KarthikGanesan88** [reply]
@beat-buesser So that seems to have fixed the variation in classification accuracy. I am now getting 94.9% classification accuracy during each run. But it is still not matching the accuracy of the PyTorch model.

**Reply by beat-buesser** [reply]
@KarthikGanesan88 I think that's to be expected because you are using a slightly different training loop, e.g. the order of samples is now constant but still different between the two scripts.

I think your training loop with the data loaders should be more efficient. I'm wondering why it is 5x slower. Are you running on GPU or CPU only?

**Reply by KarthikGanesan88** [reply]
OK. Unfortunately, I know very little about ML training; my area of work is ML inference hardware so I just take training 'as-is'. 

I am running all this on GPU.

**Reply by beat-buesser** [reply]
@KarthikGanesan88 That's ok, I notice that you are very accurate about comparing models, etc., which is great! To compare you would have to remove the shuffle operation from `fit()` to ensure that, up to numerical variation, each step of SGD goes into the same direction, based on exactly the same samples in each mini-batch, as with your script with shuffle turned off.

Did you check the GPU utilisation during both running runs? Maybe the model and data does not get transferred to the device or the device is not found at all.

I think your training loop should be as fast or faster than `fit`, I'm even thinking we should replace it with your code, but we need to find first the reason for the 5x.

**Reply by KarthikGanesan88** [reply]
> To compare you would have to remove the shuffle operation from fit()

Sorry, which `fit()` are you referring to? 

> Did you check the GPU utilisation during both running runs?

Yes, I'm using GPU in both instances. When I intentionally run it on CPU as a test, classification is noticeably slower. 

to be honest, I'm OK with this since I can train my model in PyTorch and load it in ART. But I just wanted to highlight the issue if there is some reason ART is slower as I am seeing. 

Also, my concern is if this discrepancy is due to some issue in ART which could potentially affect the results of attacks. As long as that is not the case, the variation in training between the two isn't a big issue for me.

**Reply by beat-buesser** [reply]
@KarthikGanesan88 Oh, I see, that makes sense. I thought your code was slower. The current fit loop in PyTorchClassifier is not using data loaders, for example, which is likely a source of the delays, also the shuffling should be improved.

Would you be interested to update the training loop in `PyTorchClassifier.fit` with your code?

**Reply by KarthikGanesan88** [reply]
@beat-buesser Yes, that makes sense. I don't have to manually move the data to the GPU, the dataloader handles that for me so maybe in the case of `ART` this is done for each input, which makes it slower. 

> Would you be interested to update the training loop in `PyTorchClassifier.fit` with your code?

Do you mean updating it for my own experiments? No, for now I'm OK to train the model in PyTorch and just use ART for checking attacks. Or do you mean change the code for ART? I'm really no expert with python or ART or anything like that! Like I said, my background is computer hardware so I wouldn't feel super comfortable making changes to a library like ART. I would just be worried about breaking something somewhere!

**Reply by beat-buesser** [reply]
@KarthikGanesan88 Yes, I mean as contribution in a pull request to ART. No worries, there is not chance to break anything, you code is well tested and we'll test it again. I would be happy to provide support and guide you through the steps, but only if you like and have time.

It would basically replace lines 382-414 in `PyTorchClassifier.fit()` with your lines 17-51 with some adaptations to use the provided `x` instead of loading data from disk, etc.

**Reply by KarthikGanesan88** [reply]
@beat-buesser OK. It seems like there is a robust process for git pull requests so I can give that a try! 

But I'm not sure what would be the right change(s) to make. The big thing I see is that ART uses numpy arrays throughout but PyTorch uses tensors everywhere. So when using GPU, numpy arrays would have to be copied to tensors and moved to the appropriate device. So switching things to use tensors instead should alleviate this bottleneck. But I feel like that's a much bigger change since numpy arrays are used throughout ART. 

So just to fix the speed issue with `PyTorchClassifier.fit()` I can try to convert all the data to tensors and put it on the right device at the start so it avoids any copy overhead (mainly from lines 392-393 in `PyTorchClassifier.fit()` I feel.

**Reply by beat-buesser** [reply]
@KarthikGanesan88 That's great! Let me know anytime if you have questions.

Yes, that's exactly the idea: `PyTorchClassifier.fit()` will receive data in argument `x` and optionally `y` as np.ndarray, but starting from line 382 we are free to transform them into torch.tensors, apply dataloaders, etc. as long as the changes stay inside `fit()`
